2024-08-16 18:54:33,990 - Started process with PID: 39291 on GPU: 2 | Lambda: 16, LR: 0.003, Heb_Learn: orthogonal, Heb_Growth: sigmoid, Clas_Growth: linear, Heb_Focus: neuron, Heb_Inhib: RELU, Class_Focus: neuron
2024-08-16 18:54:36,654 - Process with PID: 39291 completed on GPU: 2.
2024-08-16 18:54:36,654 - Standard Output:

2024-08-16 18:54:36,654 - Standard Error:
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
           ^^^^^^^^^^^^^^^^
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 47, in train_and_eval
    model: Network = HebbianNetwork('Hebbian Network', params).to(params.device)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/models/hebbian_network.py", line 92, in __init__
    hebbian_layer: HiddenLayer = HebbianLayer(self.input_dim,
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/layers/base/hebbian_layer.py", line 104, in __init__
    self.exponential_average: torch.Tensor = torch.zeros(self.output_dimension).to(self.device)
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tianyixu/Documents/github_projects/hebbian_venv/lib/python3.12/site-packages/torch/cuda/__init__.py", line 284, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 85, in <module>
    main()
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 22, in main
    train_acc_lists, test_acc_lists = parallel_training(ARGS, 1)
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 68, in parallel_training
    results = pool.map(train_and_eval, param_list)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/pool.py", line 774, in get
    raise self._value
AssertionError: Torch not compiled with CUDA enabled

2024-08-16 18:54:36,654 - All subprocesses have completed.
2024-08-16 18:55:34,815 - Started process with PID: 39334 on GPU: 2 | Lambda: 16, LR: 0.003, Heb_Learn: orthogonal, Heb_Growth: sigmoid, Clas_Growth: linear, Heb_Focus: neuron, Heb_Inhib: RELU, Class_Focus: neuron
2024-08-16 19:02:06,285 - Started process with PID: 39563 on GPU: 2 | Lambda: 16, LR: 0.003, Heb_Learn: orthogonal, Heb_Growth: sigmoid, Clas_Growth: linear, Heb_Focus: neuron, Heb_Inhib: RELU, Class_Focus: neuron
2024-08-16 19:05:45,261 - Started process with PID: 39697 on GPU: 2 | Lambda: 16, LR: 0.003, Heb_Learn: orthogonal, Heb_Growth: sigmoid, Clas_Growth: linear, Heb_Focus: neuron, Heb_Inhib: RELU, Class_Focus: neuron
2024-08-16 19:07:26,607 - Process with PID: 39697 completed on GPU: 2.
2024-08-16 19:07:26,607 - Standard Output:
Removed results/experiment--TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0.
Experiment -TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0 result folder re-created successfully.
Experiment '-TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0/Output' result folder re-created successfully.
Experiment '-TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0/Hidden' result folder re-created successfully.
[[0,1],[2,3],[4,5],[6,7],[8,9]]
{'MNIST_0_1': 0, 'MNIST_2_3': 0, 'MNIST_4_5': 0, 'MNIST_6_7': 0, 'MNIST_8_9': 0}
{'MNIST_0_1': 0, 'MNIST_2_3': 0, 'MNIST_4_5': 0, 'MNIST_6_7': 0, 'MNIST_8_9': 0}

2024-08-16 19:07:26,607 - Standard Error:
Traceback (most recent call last):
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 85, in <module>
    main()
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 35, in main
    avg_test = average(test_acc)
               ^^^^^^^^^^^^^^^^^
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/utils/experiment_stats.py", line 14, in average
    return round(avg, 4)
           ^^^^^^^^^^^^^
TypeError: type Tensor doesn't define __round__ method

2024-08-16 19:07:26,607 - All subprocesses have completed.
2024-08-16 19:10:03,791 - Started process with PID: 39811 on GPU: 2 | Lambda: 16, LR: 0.003, Heb_Learn: orthogonal, Heb_Growth: sigmoid, Clas_Growth: linear, Heb_Focus: neuron, Heb_Inhib: RELU, Class_Focus: neuron
2024-08-16 19:33:28,413 - Process with PID: 39811 completed on GPU: 2.
2024-08-16 19:33:28,413 - Standard Output:
Removed results/experiment--TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0.
Experiment -TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0 result folder re-created successfully.
Experiment '-TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0/Output' result folder re-created successfully.
Experiment '-TESTING-forget-0.003--orthogonal-sigmoid-neuron-relu-16.0---output_contrastive-linear-neuron-0/Hidden' result folder re-created successfully.
[[0,1],[2,3],[4,5],[6,7],[8,9]]
{'MNIST_0_1': 0, 'MNIST_2_3': 0, 'MNIST_4_5': 0, 'MNIST_6_7': 0, 'MNIST_8_9': 0}
{'MNIST_0_1': 0, 'MNIST_2_3': 0, 'MNIST_4_5': 0, 'MNIST_6_7': 0, 'MNIST_8_9': 0}

2024-08-16 19:33:28,413 - Standard Error:
Traceback (most recent call last):
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 85, in <module>
    main()
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/train_forget.py", line 35, in main
    avg_test = average(test_acc)
               ^^^^^^^^^^^^^^^^^
  File "/Users/tianyixu/Documents/github_projects/biological-deep-learning/utils/experiment_stats.py", line 14, in average
    return round(avg, 4)
           ^^^^^^^^^^^^^
TypeError: type Tensor doesn't define __round__ method

2024-08-16 19:33:28,413 - All subprocesses have completed.
